rfcs/2022-11-04-automatic-component-validation.md

```diff
@@ -0,0 +1,512 @@
+that updating it to do so in the future could potentially be hard work, as all components may need
+to be slightly refactored to deal with changes to the core traits, and so on. This is relevant to
+the design, as it stands, having no clear way to automatically derive the implementation of the
+proposed trait from existing data, forcing all implementations to be written manually.
+
+## Prior Art
+
+Vector already has existing mechanisms for validating components in terms of the Component
+Specification, and so on. This comes in the form of familiar helper methods like
+`assert_sink_compliance`, `assert_source_error`, etc. Those helpers operate with a similar approach
+to this proposal in that we run the relevant code in a way that ensures we can observe the state of
+the system before and after the code runs, calculating the validation results at the end.
+
+We do not have an existing mechanism for asserting the _correctness_ of internal telemetry, only
+that specific bits of internal telemetry are being emitted as intended. This is not necessarily a
+technical limitation, but a practical limitation of not yet having gotten to the point of adding
+support to validate the correctness.
+
+What both of these things do not provide, that this proposal does, is a mechanism to _ensure_ that a
+component that can be used has been actually tested against the intended criteria/specifications/etc
+and met them. While this doesn't happen as a result of the compiler/type system, and so is still
+somewhat fallible at the edges -- we have to actually do work to integrate this proposal with CI,
+etc -- it achieves uniform coverage across all components and provides the scaffolding to avoid the
+human fallibility aspect of "oops, we're not actually validating this component".
+
+It also heavily centralizes this validation logic such that any future work around updating what
+gets validated, or exposing more information during the validation step, and so on, can happen as a
+set of changes to a relatively small set of files, rather than, again, a change to every component
+involved, which is exactly where human fallibility has struck us in the past.
+
+## Alternatives
+
+One alternative approach could be to actually build Vector and run an entire configuration, much
+like an integration test, and we would send known data to Vector, collect the output, so on and so
+forth, much like this proposal... but we would simply collect it externally, much like `lading` does
+in our soak tests.
+
+The biggest reason not to go with this alternative approach is that it still involves much of the
+scaffolding present in this proposal, such as getting a list of all components that should be
+tested, and writing code that can emit that "external resource" information, since a Vector
+configuration may depend heavily on default values, and so we would not always be able to extract
+all of the necessary information from a Vector configuration that was otherwise valid by itself.
```

Comment:
So this alternative is roughly what I had in mind when thinking about this problem, and I'm not totally convinced (at least yet) that it would have roughly equal downsides. It certainly could be done in a way that's mostly equivalent, but it also seems to have the potential to at least start out meaningfully simpler.

For example, getting a list of all components that should be tested could just be the output of `vector list` minus some configured list of exempt components. The "external resource" information probably doesn't need to be totally automatic, and we could specify dependencies in a small metadata file per test (alongside a config snippet and input and expected output fixtures).

I'd hope that this approach could lead to:
1. Less code to write, since we don't need a new trait for components or a new way to run them
2. More confidence in the results, since Vector will be running exactly as it does for users (e.g. no concern about single-threaded Tokio, differences between reading the thread-locals and what actually gets emitted, etc)
3. Better separation of concerns, where components themselves (and the rest of the Vector codebase) don't need to know about this particular style of test

This would certainly move the initial version more towards a fixture-based integration test-style thing, but I think it could still capture the important parts of the "automatic" aspect (i.e. picking up new components by default, shared configurable output validators, etc).

Reply:
I think your proposed approach is simpler to write but more complex and thus prone to being approximately as useful as everything we currently have and everything we've tried.

- Getting the list of components to test from `vector list`: yes, simpler
- Running exactly as Vector normally does for users: sure, technically not wrong
- Better separation of concerns: sure, much more of a blackbox testing approach

And then we move into the actual things that _developers_ have to care about:

1. Every single component needs to be run as a full config (source->sink at minimum, or source->transform->sink when testing a transform) which can probably be done using a maximally-permissive filler component (i.e. if we're testing a sink, we'd always use the `vector` source to feed data, or the `vector` sink to get the data out, becuase they can handle any event type) but I haven't thought through if running a full config will carry any risk of pollution of metrics/handling/etc.
2. We have to write manual files describing how to test the component.

The first item, to me, is more of just...  haven't thought terribly deeply about if there could be any complications here. The second feels like the dealbreaker to me.

The whole entire crux of the RFC, like entirely ignoring what is actually being validated, and focusing on the mechanism that drives the tests, is that we have failed multiple times (no offense to anybody who has done this work, because I myself was responsible for one of those failed attempts) to take a human-based approach at testing these things.

Every scheme that involves a developer having to manually write some test specification that has room for fallibility -- not strongly-typed, dependent on special files existing in just the right place, and so on -- feels dead on arrival to me.

Reply:
> We have to write manual files describing how to test the component.

To be clear, my position is definitely not "just write integration tests". I'm fully in favor of making this as automated and fail-safe as possible, I just think that running in-process vs out-of-process is an orthogonal concern from that perspective. The logic is fundamentally the same, so we should implement whichever is easier.

Another aspect is that we are likely to need to configure each "test case", with external resources as the RFC stands, and probably also inputs that match each component's expectations (especially if we want to cover any failure cases). In my opinion, it'd be nice to capture that configuration as data rather than code.

I'm imagining a process roughly like the following pseudocode:
```
for component in $(vector list) {
    if component in config.excluded_components {
        continue
    }

    let scenarios = $(ls validation_tests/$config/*);
    if scenarios.is_emtpy() {
        if component.type != "transform" {
            scenarios.push(default_transform_scenario());
        } else {
            fail("component is neither validated nor explicitly excluded from validation");
        }
    }

    for scenario in $(ls validation_tests/$config/*) {
        let config = build_config(component, scenario);
        let external = spawn(scenario.external_resources);
        let output = run_and_collect_output(config, scenario.input || default_input());

        default_validator.validate(output);
        for validator in scenario.additional_validators {
            validator.validate(output);
        }
    }
}
```

This gets you a few properties that seem like we're going for here:
* You don't have to do anything for transforms to get tested
* Sources and sink are only required to specify external resources
* The default validations (whatever common bits we can require across every component) are always run
* You can explicitly specify the inputs to test with for components that need it (e.g. syslog source needs syslog input)
* You can explicitly configure additional input/validator combos to test things like failure paths

That's not to say that this version is better than the in-memory version, just that you should be able to achieve the goals of the RFC either way. I'm writing this really just to make sure that the alternative is adequately considered.

Reply:
Marking as resolved as the RFC has been significantly reworked.